### PR TITLE
Tool to close pull requests

### PR DESCRIPTION
**Pull request by apelisse**
Try an alternative approach to "closing staling PR" aka #891

### CODE REVIEW

File: close-pull-requests/.gitignore [hunk 1]
```diff
@@ -0,0 +1 @@
+close-pull-requests
```
Comment (fejta) [inline close-pull-requests/.gitignore]: Please squash this commit into another one

File: close-pull-requests/main.go [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+type closerConfig struct {
+	github.Config
+
+	warnBefore   time.Duration
+	warnReminder time.Duration
+	closeAfter   time.Duration
+	dryRun       bool
+}
+
+func addRootFlags(cmd *cobra.Command, config *closerConfig) {
```
Comment (fejta) [inline close-pull-requests/main.go]: This needs a dry-run option

File: close-pull-requests/main.go [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+	closingComment = `This PR hasn't been active in %s. Feel free to reopen.
+
+You can add 'keep-open' label to prevent this from happening again.`
+	warningComment = `This PR hasn't been active in %s. Will be closed in %s.
+
+You can add 'keep-open' label to prevent this from happening.`
+)
+
+var (
+	closingCommentRE = regexp.MustCompile(`This PR hasn't been active in \d+ days?\. Feel free to reopen.
```
Comment (fejta) [inline close-pull-requests/main.go]: I want this configurable

File: close-pull-requests/main.go [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+}
+
+func addRootFlags(cmd *cobra.Command, config *closerConfig) {
+	cmd.Flags().DurationVar(&config.warnBefore, "warn-before", 60*day, "Number of hours before we start warning")
+	cmd.Flags().DurationVar(&config.warnReminder, "warn-reminder", 30*day, "Number of hours before we warn again")
+	cmd.Flags().DurationVar(&config.closeAfter, "close-after", 90*day, "Number of hours before we actually close pull-request")
+}
+
+func isBotName(name string) bool {
+	return name == "k8s-merge-robot" || name == "k8s-bot"
```
Comment (fejta) [inline close-pull-requests/main.go]: Please make this list configurable

File: close-pull-requests/main.go [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+
+	"github.com/golang/glog"
+	githubapi "github.com/google/go-github/github"
+	"github.com/spf13/cobra"
+	"k8s.io/contrib/mungegithub/github"
+)
+
+const (
+	day            = 24 * time.Hour
+	keepOpenLabel  = "keep-open"
```
Comment (fejta) [inline close-pull-requests/main.go]: Does this label exist? It should be configurable. Also there should be a label noting when it is stale

File: close-pull-requests/main.go [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+	closingCommentRE = regexp.MustCompile(`This PR hasn't been active in \d+ days?\. Feel free to reopen.
+
+You can add 'keep-open' label to prevent this from happening again\.`)
+	warningCommentRE = regexp.MustCompile(`This PR hasn't been active in \d+ days?\. Will be closed in \d+ days?\.
+
+You can add 'keep-open' label to prevent this from happening\.`)
+)
+
+type closerConfig struct {
+	github.Config
```
Comment (fejta) [inline close-pull-requests/main.go]: What sets the repository?

File: close-pull-requests/main.go [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+	}
+	return nil
+}
+
+func runProgram(config *closerConfig) error {
+	if err := config.PreExecute(); err != nil {
+		return err
+	}
+
+	return config.ForEachIssueDo(func(obj *github.MungeObject) error {
```
Comment (fejta) [inline close-pull-requests/main.go]: See the design doc I sent which suggests a better way of doing this than iterating through every PR -- aka only search for PRs that have not been modified in a long time (to make stale/close) or issues with the stale label that have been modified recently (to remove the stale label)

**Comment by eparis** [comment]
Why is this not a mungegithub munger? We already have at least 2 other people who started this. Maybe you can take the lead on those PRs?  I'll go find them...

File: close-pull-requests/main.go [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"regexp"
+	"time"
+
+	"github.com/golang/glog"
+	githubapi "github.com/google/go-github/github"
```
Comment (fejta) [inline close-pull-requests/main.go]: nit: I'm not entirely convinced that all the gain from all these dependencies is worth the costs.

**Comment by eparis** [comment]
wait, one of those PRs is yours!
https://github.com/kubernetes/contrib/pull/891
https://github.com/kubernetes/contrib/pull/1176

**Comment by bgrant0607** [comment]
@apelisse  Please don't create PRs with no description. What is this?

**Comment by apelisse** [comment]
Sorry, I've created that in a hurry. As discussed earlier, I'd rather get #891 merged instead and close this.

**Pull request closed without merging**